File: hash/max_fruits.py

def harvest_max_fruits(fruits: list[list[int]], start_position: int, steps: int) -> int:  # LeetCode Q.2106.
    positions2fruits = {position: fruit for position, fruit in fruits}

    route_1_min_position = start_position - steps  # Route 1: go right and turn left.
    route_1_max_position = start_position

    route_1_fruits = 0
    for position in range(start_position - steps, start_position + 1):
        if position in positions2fruits.keys():
            route_1_fruits += positions2fruits[position]

    route_2_min_position = start_position  # Route 2: go left and turn right.
    route_2_max_position = start_position + steps

    route_2_fruits = 0
    for position in range(start_position, start_position + steps + 1):
        if position in positions2fruits.keys():
            route_2_fruits += positions2fruits[position]

    max_fruits = max(route_1_fruits, route_2_fruits)

    max_pre_turn_len = (steps + 1) // 2 - 1
    for _ in range(max_pre_turn_len):
        route_1_min_position += 2
        route_1_max_position += 1

        for position in (route_1_min_position - 2, route_1_min_position - 1):
            if position in positions2fruits.keys():
                route_1_fruits -= positions2fruits[position]

        if route_1_max_position in positions2fruits.keys():
            route_1_fruits += positions2fruits[route_1_max_position]

        route_2_min_position -= 1
        route_2_max_position -= 2

        for position in (route_2_max_position + 1, route_2_max_position + 2):
            if position in positions2fruits.keys():
                route_2_fruits -= positions2fruits[position]

        if route_2_min_position in positions2fruits.keys():
            route_2_fruits += positions2fruits[route_2_min_position]

        if max(route_1_fruits, route_2_fruits) > max_fruits:
            max_fruits = max(route_1_fruits, route_2_fruits)

    return max_fruits

File: hash/max_fruits/max_fruits.cpp
#include <vector>
#include <algorithm>
using namespace std;

int harvestMaxFruits(vector<vector<int>> &fruits, int startPosition, int steps)
{ // LeetCode Q.2106.
    vector<int> positions, fruitsPrefixSums;
    for (auto fruit : fruits) // Format: {position, fruits}.
    {
        positions.push_back(fruit[0]);

        if (fruitsPrefixSums.empty())
            fruitsPrefixSums.push_back(fruit[1]);

        else
            fruitsPrefixSums.push_back(fruitsPrefixSums.back() + fruit[1]);
    }

    int maxHarvestedFruits = 0;

    // Idx of the last position <= start position.
    int startIdx = upper_bound(positions.begin(), positions.end(), startPosition) - positions.begin() - 1;

    for (int step_1 = 0; step_1 <= steps / 2; step_1++)
    {
        int step_2 = steps - step_1 * 2;

        vector<pair<int, int>> ends_pairs = {
            {startPosition - step_1, startPosition + step_2},
            {startPosition - step_2, startPosition + step_1}};

        for (auto ends_pair : ends_pairs)
        {
            auto [left_end, right_end] = ends_pair;

            // Idx of the last position <= left end - 1.
            int leftIdx = upper_bound(positions.begin(), positions.end(), left_end - 1) - positions.begin() - 1;

            // Idx of the last position <= right end.
            int rightIdx = upper_bound(positions.begin(), positions.end(), right_end) - positions.begin() - 1;

            int harvestedFruits = 0;

            if (rightIdx > startIdx)
            { // Can harvest fruits at start position's right side.
                harvestedFruits += fruitsPrefixSums[rightIdx];
                if (startIdx >= 0)
                    harvestedFruits -= fruitsPrefixSums[startIdx];
            }

            if (startIdx > leftIdx)
            { // Can harvest fruits at start position's left side.
                harvestedFruits += fruitsPrefixSums[startIdx];
                if (leftIdx >= 0)
                    harvestedFruits -= fruitsPrefixSums[leftIdx];
            }

            if (harvestedFruits > maxHarvestedFruits)
                maxHarvestedFruits = harvestedFruits;
        }
    }

    return maxHarvestedFruits;
}
